src/models.py

```diff
@@ -45,38 +98,78 @@ def __init__(self, num_hidden, embedding_dim, vocab_size, device, num_layers=2,
         self.device = device
         
         self.emb = nn.Embedding(vocab_size, embedding_dim)
-        self.lstm = nn.LSTM(embedding_dim, num_hidden, num_layers=2, batch_first=True, dropout=0, bidirectional=bidirectional)
-        self.fc = nn.Linear(num_hidden, vocab_size)
+        self.lstm = nn.LSTM(embedding_dim, num_hidden, num_layers=2, batch_first=True, dropout=0, bidirectional=bidirectional) 
+        self.attn = nn.Linear(self.num_hidden * 2, 1) 
```

Comment:
It's unfortunate that in the Show, Attend and Tell paper they say that they use a multilayer perceptron for the attention function, but they don't really specify the details, e.g. the amount of hidden layers. My intuition is that we might want to turn this into two linear layers chained together. The reason why I think this is that the attention mechanism should model some sort of relationship between the two stacked input vectors (the decoder hidden state and one of the annotation vectors). However, in a single linear layer you cannot consider the relationship between different values of the input, which _would_ be possible if you add another linear layer after it. 

The same thing goes for the MLP class at the beginning of this file. It's unclear how many layers it should have. My proposal would be to also add a second layer there, and see how well it works.